File: src/main/java/com/pyamsoft/pydroid/behavior/HideScrollFABBehavior.java
/*
 * Copyright 2016 Peter Kenji Yamanaka
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.pyamsoft.pydroid.behavior;

import android.support.design.widget.CoordinatorLayout;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.view.ViewCompat;
import android.view.View;

public class HideScrollFABBehavior extends FloatingActionButton.Behavior {

  final int distanceNeeded;
  boolean animating = false;

  public HideScrollFABBehavior() {
    this(0);
  }

  public HideScrollFABBehavior(final int distance) {
    super();
    distanceNeeded = distance;
    animating = false;
  }

  public boolean isAnimating() {
    return animating;
  }

  public void endAnimation() {
    this.animating = false;
  }

  public void onHiddenHook() {

  }

  public void onShownHook() {

  }

  @Override
  public void onNestedScroll(CoordinatorLayout coordinatorLayout, FloatingActionButton child,
      View target, int dxConsumed, int dyConsumed, int dxUnconsumed, int dyUnconsumed) {
    super.onNestedScroll(coordinatorLayout, child, target, dxConsumed, dyConsumed, dxUnconsumed,
        dyUnconsumed);
    if (dyConsumed > distanceNeeded && child.isShown()) {
      if (!animating) {
        animating = true;
        child.hide(new FloatingActionButton.OnVisibilityChangedListener() {
          @Override public void onHidden(FloatingActionButton fab) {
            super.onHidden(fab);
            onHiddenHook();
            animating = false;
          }
        });
      }
    } else if (dyConsumed < -distanceNeeded && !child.isShown()) {
      if (!animating) {
        animating = true;
        child.show(new FloatingActionButton.OnVisibilityChangedListener() {
          @Override public void onShown(FloatingActionButton fab) {
            super.onShown(fab);
            onShownHook();
            animating = false;
          }
        });
      }
    }
  }

  @Override public boolean onStartNestedScroll(CoordinatorLayout coordinatorLayout,
      FloatingActionButton child, View directTargetChild, View target, int nestedScrollAxes) {
    return nestedScrollAxes == ViewCompat.SCROLL_AXIS_VERTICAL;
  }
}

File: pydroid-design/src/main/java/com/pyamsoft/pydroid/design/fab/HideScrollFABBehavior.kt
/*
 * Copyright (C) 2018 Peter Kenji Yamanaka
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.pyamsoft.pydroid.design.fab

import android.support.annotation.CheckResult
import android.support.design.widget.CoordinatorLayout
import android.support.design.widget.FloatingActionButton
import android.support.v4.view.ViewCompat
import android.view.View
import androidx.view.isInvisible
import androidx.view.isVisible
import com.pyamsoft.pydroid.design.util.hide
import com.pyamsoft.pydroid.design.util.show
import timber.log.Timber

/**
 * Floating Action Button behavior which hides button after scroll distance is passed
 */
class HideScrollFABBehavior(
  private val distanceNeeded: Int = 0,
  private val onHidden: (FloatingActionButton) -> Unit = {},
  private val onShown: (FloatingActionButton) -> Unit = {}
) : FloatingActionButton.Behavior() {

  var isAnimating = false
    @get:CheckResult get
    private set

  fun endAnimation() {
    this.isAnimating = false
  }

  private fun hideFab(fab: FloatingActionButton) {
    isAnimating = true
    Timber.w("Hide FAB")
    fab.hide {
      // Support library as on 25.1.0 sets FAB visibility to GONE
      // making it ignore other scrolling event.
      // Set it to invisible to fix this problem
      isInvisible = true
      isAnimating = false
      onHidden(this)
    }
  }

  private fun showFab(fab: FloatingActionButton) {
    isAnimating = true
    Timber.w("Show FAB")
    fab.show {
      isVisible  = true
      isAnimating = false
      onShown(this)
    }
  }

  override fun onNestedScroll(
    coordinatorLayout: CoordinatorLayout,
    child: FloatingActionButton,
    target: View,
    dxConsumed: Int,
    dyConsumed: Int,
    dxUnconsumed: Int,
    dyUnconsumed: Int,
    type: Int
  ) {
    if (dyConsumed > distanceNeeded && child.isShown) {
      if (!isAnimating) {
        hideFab(child)
      }
    } else if (dyConsumed < -distanceNeeded && !child.isShown) {
      if (!isAnimating) {
        showFab(child)
      }
    }
  }

  override fun onStartNestedScroll(
    coordinatorLayout: CoordinatorLayout,
    child: FloatingActionButton,
    directTargetChild: View,
    target: View,
    axes: Int,
    type: Int
  ): Boolean = (axes == ViewCompat.SCROLL_AXIS_VERTICAL)
}
